relay/writer/file.go

```diff
@@ -0,0 +1,381 @@
+		err := w.out.Close()
+		if err != nil {
+			return nil, errors.Annotate(err, "close previous underlying binlog writer")
+		}
+	}
+
+	// open/create a new binlog file
+	filename := filepath.Join(w.cfg.RelayDir, w.filename.Get())
+	outCfg := &bw.FileWriterConfig{
+		Filename: filename,
+	}
+	out := bw.NewFileWriter(outCfg)
+	err := out.Start()
+	if err != nil {
+		return nil, errors.Annotatef(err, "start underlying binlog writer for %s", filename)
+	}
+	w.out = out
+	log.Infof("[relay] open underlying binlog writer with status %v", w.out.Status())
+
+	// write the binlog file header if not exists
+	exist, err := checkBinlogHeaderExist(filename)
+	if err != nil {
+		return nil, errors.Annotatef(err, "check binlog file header for %s", filename)
+	} else if !exist {
+		err = w.out.Write(replication.BinLogFileHeader)
+		if err != nil {
+			return nil, errors.Annotatef(err, "write binlog file header for %s", filename)
+		}
+	}
+
+	// write the FormatDescriptionEvent if not exists one
+	exist, err = checkFormatDescriptionEventExist(filename)
+	if err != nil {
+		return nil, errors.Annotatef(err, "check FormatDescriptionEvent for %s", filename)
+	} else if !exist {
+		err = w.out.Write(ev.RawData)
+		if err != nil {
+			return nil, errors.Annotatef(err, "write FormatDescriptionEvent %+v for %s", ev.Header, filename)
+		}
+	}
+
+	// record the offset of the opened/created binlog file
+	outF, ok := w.out.(*bw.FileWriter)
+	if !ok {
+		return nil, errors.New("underlying binlog writer must be a FileWriter now")
+	}
+	outFS := outF.Status().(*bw.FileWriterStatus)
+	w.offset.Set(outFS.Offset)
+
+	return &Result{
+		Ignore: exist, // ignore if exists
+	}, nil
+}
+
+// handle RotateEvent:
+//   1. update binlog filename if needed
+//   2. write the RotateEvent if not fake
+//   3. update offset if the event is wrote
+// NOTE: we do not create a new binlog file when received a RotateEvent,
+//       instead, we create a new binlog file when received a FormatDescriptionEvent.
+//       because a binlog file without any events has no meaning.
+func (w *FileWriter) handleRotateEvent(ev *replication.BinlogEvent) (*Result, error) {
+	rotateEv, ok := ev.Event.(*replication.RotateEvent)
+	if !ok {
+		return nil, errors.NotValidf("except RotateEvent, but got %+v", ev.Header)
+	}
+
+	// update binlog filename if needed
+	var currFile = w.filename.Get()
+	nextFile := string(rotateEv.NextLogName)
+	if nextFile > currFile {
+		// record the next filename, but not create it.
+		// even it's a fake RotateEvent, we still need to record it,
+		// because if we do not specify the filename when creating the writer (like Auto-Position),
+		// we can only receive a fake RotateEvent before the FormatDescriptionEvent.
+		w.filename.Set(nextFile)
+	}
+
+	// write the RotateEvent if not fake
+	var ignore bool
+	if ev.Header.Timestamp == 0 || ev.Header.LogPos == 0 {
+		// skip fake rotate event
+		ignore = true
+	} else if w.out == nil {
```

Comment:
must it open `out` at `handleFormatDescriptionEvent`?

Reply:
yes, before any non-fake events (needed to be written) are sent, a valid `FormatDescriptionEvent` event will always be sent, so we can open `out` at `handleFormatDescriptionEvent`